File: go/0494M-目标和.go
/*
 * @Author: SukiEva
 * @Date: 2021-12-01 09:09:09
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

import "math"

func findTargetSumWays(nums []int, target int) int {
	sum := 0
	for _, num := range nums {
		sum += num
	}
	if math.Abs(float64(target)) > float64(sum) || (target+sum)&1 == 1 {
		return 0
	}
	sum = (target + sum) >> 1
	dp := make([]int, sum+1)
	dp[0] = 1
	for _, num := range nums {
		for j := sum; j >= num; j-- {
			dp[j] += dp[j-num]
		}
	}
	return dp[sum]
}

File: cpp/0494M-目标和.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-01-10 12:57:29
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

/*
l - r = target
l + r = sum
==> l = (target+sum)/2
==> 目标和为 l 有多少种
*/

class Solution {
   public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int sum = 0;
        for (int num : nums) sum += num;
        if (abs(target) > sum || (sum + target) & 1) return 0;
        target = (sum + target) >> 1;
        vector<int> dp(target + 1);
        dp[0] = 1;
        for (int i = 0; i < nums.size(); i++) {
            for (int j = target; j >= nums[i]; j--) {
                dp[j] += dp[j - nums[i]];
            }
        }
        return dp[target];
    }
};